lib/Scheduling/CPSATSchedulers.cpp

```diff
@@ -0,0 +1,138 @@
+//===- CPSATSchedulers.cpp - Schedulers using external CPSAT solvers
+//------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of cp-sat programming-based schedulers using external solvers
+// via OR-Tools.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Scheduling/Algorithms.h"
+
+#include "mlir/IR/Operation.h"
+
+#include "ortools/sat/cp_model.h"
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Format.h"
+
+#define DEBUG_TYPE "cpsat-schedulers"
+
+using namespace circt;
+using namespace circt::scheduling;
+using namespace operations_research;
+using namespace operations_research::sat;
+
+using llvm::dbgs;
+using llvm::format;
+
+LogicalResult scheduling::scheduleCPSAT(SharedOperatorsProblem &prob,
+                                        Operation *lastOp) {
+  Operation *containingOp = prob.getContainingOp();
+  if (!prob.hasOperation(lastOp))
+    return containingOp->emitError("problem does not include last operation");
+
+  CpModelBuilder cpModel;
+  auto &tasks = prob.getOperations();
+  auto &resources = prob.getOperatorTypes();
+
+  DenseMap<Operation *, IntVar> taskStarts;
+  DenseMap<Operation *, IntVar> taskEnds;
+  DenseMap<Problem::OperatorType, llvm::SmallVector<IntervalVar, 4>>
+      resourcesToTaskIntervals;
+
+  // first get horizon
+  unsigned horizon = 0;
+  for (auto *task : tasks) {
+    unsigned duration = *prob.getLatency(*prob.getLinkedOperatorType(task));
+    horizon += duration;
+  }
+
+  // build vars
+  for (auto item : llvm::enumerate(tasks)) {
+    auto i = item.index();
+    auto *task = item.value();
+    IntVar startVar = cpModel.NewIntVar(Domain(0, horizon))
+                          .WithName((Twine("start_of_task_") + Twine(i)).str());
+    IntVar endVar = cpModel.NewIntVar(Domain(0, horizon))
+                        .WithName((Twine("end_of_task_") + Twine(i)).str());
+    taskStarts[task] = startVar;
+    taskEnds[task] = endVar;
+    auto resource = prob.getLinkedOperatorType(task);
+    if (resource.hasValue()) {
+      unsigned duration = *prob.getLatency(*resource);
+      IntervalVar taskInterval =
+          cpModel.NewIntervalVar(startVar, duration, endVar)
+              .WithName((Twine("task_interval_") + Twine(i)).str());
+
+      if (!resourcesToTaskIntervals.count(resource.value()))
+        resourcesToTaskIntervals[resource.value()] = {};
```

Comment:
Interesting. Am I just misinformed or do `stl` maps/sets behave the same way (i.e. I've never needed to do this and somehow learned a weird habit)?

Reply:
[std::map](https://en.cppreference.com/w/cpp/container/map/operator_at):
> operator[] is non-const because it inserts the key if it doesn't exist. If this behavior is undesirable or if the container is const, [at()](https://en.cppreference.com/w/cpp/container/map/at) may be used.